File: heron/instance/src/java/org/apache/heron/instance/bolt/BoltOutputCollectorImpl.java
/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package org.apache.heron.instance.bolt;

import java.time.Duration;
import java.util.Collection;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.logging.Level;
import java.util.logging.Logger;

import com.google.protobuf.Message;

import org.apache.heron.api.bolt.IOutputCollector;
import org.apache.heron.api.serializer.IPluggableSerializer;
import org.apache.heron.api.tuple.Tuple;
import org.apache.heron.common.basics.Communicator;
import org.apache.heron.common.utils.metrics.IBoltMetrics;
import org.apache.heron.common.utils.misc.PhysicalPlanHelper;
import org.apache.heron.common.utils.tuple.TupleImpl;
import org.apache.heron.instance.AbstractOutputCollector;
import org.apache.heron.proto.system.HeronTuples;

/**
 * BoltOutputCollectorImpl is used by bolt to emit tuples, it contains:
 * 1. IPluggableSerializer serializer, which will define the serializer
 * 2. OutgoingTupleCollection outputter.
 * When a tuple is to be emitted, it will serialize it and call OutgoingTupleCollection.admitBoltTuple()
 * to sent it out.
 * <p>
 * It will handle the extra work to emit a tuple:
 * For data tuples:
 * 1. Set the anchors for a tuple
 * 2. Pack the tuple and submit the OutgoingTupleCollection's addDataTuple
 * 3. Update the metrics
 * <p>
 * For Control tuples (ack &amp; fail):
 * 1. Set the anchors for a tuple
 * 2. Pack the tuple and submit the OutgoingTupleCollection's addDataTuple
 * 3. Update the metrics
 */
public class BoltOutputCollectorImpl extends AbstractOutputCollector implements IOutputCollector {
  private static final Logger LOG = Logger.getLogger(BoltOutputCollectorImpl.class.getName());

  // Reference to update the bolt metrics
  private final IBoltMetrics boltMetrics;

  protected BoltOutputCollectorImpl(IPluggableSerializer serializer,
                                    PhysicalPlanHelper helper,
                                    Communicator<Message> streamOutQueue,
                                    IBoltMetrics boltMetrics) {
    super(serializer, helper, streamOutQueue, boltMetrics);
    this.boltMetrics = boltMetrics;

    if (helper.getMyBolt() == null) {
      throw new RuntimeException(helper.getMyTaskId() + " is not a bolt ");
    }
  }

  @Override
  public List<Integer> emit(String streamId, Collection<Tuple> anchors, List<Object> tuple) {
    return admitBoltTuple(streamId, anchors, tuple, null);
  }

  @Override
  public void emitDirect(int taskId, String streamId,
                         Collection<Tuple> anchors, List<Object> tuple) {
    admitBoltTuple(streamId, anchors, tuple, taskId);
  }

  @Override
  public void reportError(Throwable error) {
    LOG.log(Level.SEVERE, "Reporting an error in topology code ", error);
  }

  @Override
  public void ack(Tuple input) {
    admitAckTuple(input);
  }

  @Override
  public void fail(Tuple input) {
    admitFailTuple(input);
  }

  /////////////////////////////////////////////////////////
  // Following private methods are internal implementations
  /////////////////////////////////////////////////////////
  private List<Integer> admitBoltTuple(String streamId,
                                       Collection<Tuple> anchors,
                                       List<Object> tuple,
                                       Integer emitDirectTaskId) {
    if (getPhysicalPlanHelper().isTerminatedComponent()) {
      // No need to handle this tuple
      return null;
    }

    // Start construct the data tuple
    HeronTuples.HeronDataTuple.Builder bldr = initTupleBuilder(streamId, tuple, emitDirectTaskId);

    // Set the anchors for a tuple
    if (anchors != null) {
      // This message is rooted
      Set<HeronTuples.RootId> mergedRoots = new HashSet<>();
      for (Tuple tpl : anchors) {
        if (tpl instanceof TupleImpl) {
          TupleImpl t = (TupleImpl) tpl;
          mergedRoots.addAll(t.getRoots());
        }
      }
      for (HeronTuples.RootId rt : mergedRoots) {
        bldr.addRoots(rt);
      }
    }

    sendTuple(bldr, streamId, tuple);

    // TODO:- remove this after changing the API
    return null;
  }

  private void admitAckTuple(Tuple tuple) {
    Duration latency = Duration.ZERO;
    if (ackEnabled) {
      if (tuple instanceof TupleImpl) {
        TupleImpl tuplImpl = (TupleImpl) tuple;

        HeronTuples.AckTuple.Builder bldr = HeronTuples.AckTuple.newBuilder();
        bldr.setAckedtuple(tuplImpl.getTupleKey());

        long tupleSizeInBytes = 0;

        for (HeronTuples.RootId rt : tuplImpl.getRoots()) {
          bldr.addRoots(rt);
          tupleSizeInBytes += rt.getSerializedSize();
        }
        outputter.addAckTuple(bldr, tupleSizeInBytes);

        latency = Duration.ofNanos(System.nanoTime()).minusNanos(tuplImpl.getCreationTime());
      }
    }

    // Invoke user-defined boltAck task hook
    getPhysicalPlanHelper().getTopologyContext().invokeHookBoltAck(tuple, latency);

    boltMetrics.ackedTuple(
        tuple.getSourceStreamId(), tuple.getSourceComponent(), latency.toNanos());
  }

  private void admitFailTuple(Tuple tuple) {
    Duration latency = Duration.ZERO;
    if (ackEnabled) {
      if (tuple instanceof TupleImpl) {
        TupleImpl tuplImpl = (TupleImpl) tuple;

        HeronTuples.AckTuple.Builder bldr = HeronTuples.AckTuple.newBuilder();
        bldr.setAckedtuple(tuplImpl.getTupleKey());

        long tupleSizeInBytes = 0;

        for (HeronTuples.RootId rt : tuplImpl.getRoots()) {
          bldr.addRoots(rt);
          tupleSizeInBytes += rt.getSerializedSize();
        }
        outputter.addFailTuple(bldr, tupleSizeInBytes);

        latency = Duration.ofNanos(System.nanoTime()).minusNanos(tuplImpl.getCreationTime());
      }
    }

    // Invoke user-defined boltFail task hook
    getPhysicalPlanHelper().getTopologyContext().invokeHookBoltFail(tuple, latency);

    boltMetrics.failedTuple(
        tuple.getSourceStreamId(), tuple.getSourceComponent(), latency.toNanos());
  }
}

File: heron/instance/src/cpp/boltimpl/bolt-output-collector-impl.cpp
/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

#include <list>
#include <map>
#include <set>
#include <string>
#include <vector>

#include "boltimpl/bolt-output-collector-impl.h"
#include "proto/messages.h"
#include "network/network.h"
#include "basics/basics.h"

#include "boltimpl/tuple-impl.h"
#include "boltimpl/tick-tuple.h"

struct rootCompare {
  bool operator() (const heron::proto::system::RootId& lhs,
                   const heron::proto::system::RootId& rhs) const {
    return lhs.taskid() < rhs.taskid() ||
           (lhs.taskid() == rhs.taskid() && lhs.key() < rhs.key());
  }
};

namespace heron {
namespace instance {

BoltOutputCollectorImpl::BoltOutputCollectorImpl(
                          std::shared_ptr<api::serializer::IPluggableSerializer> serializer,
                          std::shared_ptr<TaskContextImpl> taskContext,
                          NotifyingCommunicator<google::protobuf::Message*>* dataFromExecutor,
                          std::shared_ptr<BoltMetrics> metrics)
  : api::bolt::IBoltOutputCollector(serializer), metrics_(metrics) {
  collector_ = new OutgoingTupleCollection(taskContext->getThisComponentName(), dataFromExecutor);
  ackingEnabled_ = taskContext->isAckingEnabled();
  taskId_ = taskContext->getThisTaskId();
}

BoltOutputCollectorImpl::~BoltOutputCollectorImpl() {
  delete collector_;
}

void BoltOutputCollectorImpl::reportError(std::exception& except) {
  LOG(INFO) << "Reporting an error in topology code " << except.what();
}

void BoltOutputCollectorImpl::emitInternal(const std::string& streamId,
                                           std::list<std::shared_ptr<api::tuple::Tuple>>& anchors,
                                           const std::vector<std::string>& tup) {
  auto msg = new proto::system::HeronDataTuple();
  msg->set_key(0);
  std::set<proto::system::RootId, rootCompare> mergedRoots;
  for (auto anchor : anchors) {
    std::shared_ptr<TupleImpl> t = std::dynamic_pointer_cast<TupleImpl>(anchor);
    if (t) {
      std::shared_ptr<const proto::system::HeronDataTuple> actualRepr = t->getDataTuple();
      for (int i = 0; i < actualRepr->roots_size(); ++i) {
        mergedRoots.insert(actualRepr->roots(i));
      }
    }
  }
  for (auto root : mergedRoots) {
    msg->add_roots()->CopyFrom(root);
  }
  int64_t totalSize = 0;
  for (auto& s : tup) {
    totalSize += s.size();
  }
  collector_->addDataTuple(streamId, msg, totalSize);
  metrics_->emittedTuple(streamId);
}

void BoltOutputCollectorImpl::ack(std::shared_ptr<api::tuple::Tuple> tup) {
  if (ackingEnabled_) {
    std::shared_ptr<TupleImpl> t = std::dynamic_pointer_cast<TupleImpl>(tup);
    if (t) {
      proto::system::AckTuple* ack = new proto::system::AckTuple();
      std::shared_ptr<const proto::system::HeronDataTuple> actualRepr = t->getDataTuple();
      ack->set_ackedtuple(actualRepr->key());
      int64_t tupSize = 0;
      for (int i = 0; i < actualRepr->roots_size(); ++i) {
        ack->add_roots()->CopyFrom(actualRepr->roots(i));
        tupSize += actualRepr->roots(i).ByteSizeLong();
      }
      collector_->addAckTuple(ack, tupSize);
      int64_t currentTime = std::chrono::duration_cast<std::chrono::nanoseconds>(
                                   std::chrono::system_clock::now().time_since_epoch()).count();
      metrics_->ackedTuple(t->getSourceStreamId(), t->getSourceComponent(),
                           currentTime - t->getCreationTimeNs());
    }
  }
}

void BoltOutputCollectorImpl::fail(std::shared_ptr<api::tuple::Tuple> tup) {
  if (ackingEnabled_) {
    std::shared_ptr<TupleImpl> t = std::dynamic_pointer_cast<TupleImpl>(tup);
    if (t) {
      proto::system::AckTuple* fl = new proto::system::AckTuple();
      std::shared_ptr<const proto::system::HeronDataTuple> actualRepr = t->getDataTuple();
      fl->set_ackedtuple(actualRepr->key());
      int64_t tupSize = 0;
      for (int i = 0; i < actualRepr->roots_size(); ++i) {
        fl->add_roots()->CopyFrom(actualRepr->roots(i));
        tupSize += actualRepr->roots(i).ByteSizeLong();
      }
      collector_->addFailTuple(fl, tupSize);
      int64_t currentTime = std::chrono::duration_cast<std::chrono::nanoseconds>(
                                   std::chrono::system_clock::now().time_since_epoch()).count();
      metrics_->failedTuple(t->getSourceStreamId(), t->getSourceComponent(),
                            currentTime - t->getCreationTimeNs());
    }
  }
}

}  // namespace instance
}  // namespace heron
